sbncode/EventGenerator/MeVPrtl/Tools/NuMiKaonGen_tool.cc

```diff
@@ -0,0 +1,330 @@
+/**
+ *
+ */
+
+// Framework Includes
+#include "art/Framework/Core/EDProducer.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Services/Registry/ServiceHandle.h"
+#include "art/Persistency/Common/PtrMaker.h"
+#include "art/Utilities/ToolMacros.h"
+#include "cetlib/cpu_timer.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+#include "CLHEP/Random/RandFlat.h"
+#include "IFDH_service.h"
+
+#include "nusimdata/SimulationBase/MCFlux.h"
+
+// local includes
+#include "IMesonGen.h"
+
+// LArSoft includes
+#include "dk2nu/tree/dk2nu.h"
+#include "dk2nu/tree/dkmeta.h"
+#include "dk2nu/tree/NuChoice.h"
+
+// ROOT
+#include "TVector3.h"
+#include "TTree.h"
+#include "TFile.h"
+#include "TTreeReader.h"
+#include "TTreeReaderValue.h"
+
+// std includes
+#include <string>
+#include <iostream>
+#include <memory>
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+// implementation follows
+
+namespace evgen {
+namespace ldm {
+/**
+ *  @brief  NuMiKaonGen class definiton
+ */
+class NuMiKaonGen : public IMesonGen
+{
+public:
+    /**
+     *  @brief  Constructor
+     */
+    NuMiKaonGen(fhicl::ParameterSet const &pset);
+
+    /**
+     *  @brief  Destructor
+     */
+    ~NuMiKaonGen();
+
+    double GetPOT() override;
+    simb::MCFlux GetNext() override;
+
+    void configure(const fhicl::ParameterSet&) override;
+
+    const bsim::Dk2Nu *GetNextEntry();
+    std::vector<std::string> LoadFluxFiles();
+    simb::MCFlux MakeMCFlux(const bsim::Dk2Nu &dk2nu);
+    double LoadPOT();
+
+    // no weights
+    float MaxWeight() override { return -1.; }
+
+private:
+  // config
+  std::string fSearchPath;
+  std::vector<std::string> fSearchPatterns;
+  unsigned long fMaxFluxFileMB;
+  std::string fFluxCopyMethod;
+  bool fRandomizeFiles;
+
+  std::string fTreeName;
+  std::string fMetaTreeName;
+
+  // info for tracking files
+  unsigned fFileIndex;
+  bool fNewFile;
+  std::vector<std::string> fFluxFiles;
+
+  // info for tracking entry in file
+  unsigned fEntry;
+  unsigned fEntryStart;
+
+  // ROOT Holders
+  TTree *fFluxTree;
```

Comment:
I don't see what's wrong with normal pointers.

Also, for the TTree, I don't think it's possible to use a smart pointer since the TTree object is "owned" by ROOT, not by this class.

Reply:
Hi @gputnam 
You have kind of demonstrated the exact need for smart pointers above; you allocated stuff on the heap and never deleted it.  Smart ptrs are a bit fiddly but they are a practice worth investing time in.

FYI you --can-- hodgepodge root objects into smart pointers but it is faffy.  Provided that you're using the TFileService to control the TTree then I think it's OK to let the ART framework handle the I/O and object lifetime.

Reply:
Minor correction, delete or smart pointers are only necessary for the TFile, the TTree is ok as it is owned by the TFile.

Reply:
A while back I figured out how to to smart pointers for ROOT objects. It not that hard. 

My example is for a `TH1D` on the flash match. 
```
std::unique_ptr<TH1D> fOpHitsTimeHist;
fOpHitsTimeHist = std::make_unique<TH1D>("fOpHitsTimeHist", "ophittime", fTimeBins, fBeamWindowStart, fBeamWindowEnd); // in us
fOpHitsTimeHist->SetOption("HIST");
fOpHitsTimeHist->SetDirectory(0);//turn off ROOT's object ownership
```